docs/explanation/charm-architecture.md

```diff
@@ -0,0 +1,115 @@
+# Charm architecture
+
+The GitHub Runner Image Builder is a machine charm responsible for managing an application ("image-builder")
+that creates and stores VM images suitable for use by self-hosted GitHub Runners using Openstack. 
+The image-builder source code is hosted at https://github.com/canonical/github-runner-image-builder.
+
+The image-builder uses an Openstack cloud to build images. 
+Through integration with another Charm, it obtains the credentials to upload the images to a specified Openstack project,
+which can then be reused by the other Charm to spawn VM instances with the necessary software preinstalled.
+
+```mermaid
+
+C4Container
+title Container diagram for Image Builder System
+
+
+ Container_Boundary(c1, "Image Builder") {
+    Container(imagebuilder, "Image Builder", "", "Provides images to all related charms")
+ }
+    System_Ext(osbuilding, "OpenStack", "OpenStack deployment used for building images")
+
+Container_Boundary(c2, "GitHub Runner"){
+
+    Container(githubrunner, "GitHub Runner Charm", "", "Manages self-hosted runners")
+}
+    Rel(imagebuilder, osbuilding, "builds images")
+    UpdateRelStyle(imagebuilder, osbuilding, $offsetY="-30", $offsetX="10")
+    Rel(imagebuilder, osgithubrunner, "uploads images")
+    UpdateRelStyle(imagebuilder, osgithubrunner, $offsetY="-30", $offsetX="-90")
+
+
+
+
+    Rel(imagebuilder, githubrunner, "image ids")
+    UpdateRelStyle(imagebuilder, githubrunner, $offsetY="-10", $offsetX="-30")
+
+    System_Ext(osgithubrunner, "OpenStack", "OpenStack deployment used for spawning runner VMs")
+    Rel(githubrunner, osgithubrunner, "spawns VMs")
+    UpdateRelStyle(githubrunner, osgithubrunner, $offsetY="-30", $offsetX="10")
+
+    Rel(githubrunner, imagebuilder, "OpenStack credentials")
+    UpdateRelStyle(githubrunner, imagebuilder, $offsetY="10", $offsetX="-60")
+```
+
+
+The image-builder uses the [Openstack SDK](https://docs.openstack.org/openstacksdk/latest/)  to spawn a VM instance in a cloud specified
+by a config option. Using an external OpenStack VM instead of the charms machine allows for more features
+(using chroot has some limitations, e.g. for building snaps) and parallel image building.
+[cloud-init](https://cloud-init.io/) is used to install the necessary dependencies for spawning self-hosted runners
+([github actions runner binary](https://github.com/actions/runner)) and e.g. tools for automatic proxy support ([aproxy](https://github.com/canonical/aproxy)). 
+In addition, software that is heavily used in Canonical projects (such as [juju](https://juju.is/) or [microk8s](https://microk8s.io/)) 
+is pre-installed to reduce the actual CI runtime (as users do not have to install this software in their respective CI runs). 
```

Comment:
perhaps it might be more accurate to say
```suggestion
can be configured to be pre-installed to reduce the actual CI runtime (as users do not have to install this software in their respective CI runs). 
```